### ISSUE TITLE
Report view

### ISSUE THREAD

**Issue by alxngyn** [osuosl/iam#71]
I thought I would get started with views while everyone is working on backend stuff.
My initial sketch is up here: https://app.moqups.com/pvector@osuosl.org/OGiRSmIKZm/view/page/a5ffddc6e
(node 3 is clickable)

### Tech:
* jQuery
* AJAX
* D3
* http://metricsgraphicsjs.org/ (d3 plugin,  used for linked graphs)
* hosted on `GET` /report  by `app.rb`

### API
* endpoints needed:
 * clients/nodes/projects get all 
   * e.g. `/nodes` returns a json list of all available fqdn nodes
 * specific data endpoints for each plugin
   * e.g. `/DiskSize/facebook.com` returns a json with the time based data for disk sizes for node facebook.com
 * specific data endpoints for each plugin with a time parameter
   * e.g. `/DiskSize/facebook.com/15` returns a json with the time based data for disk sizes for node facebook.com for the last 15 days

### Styling
* http://v4-alpha.getbootstrap.com/ (in alpha though)
 or  https://www.google.com/design/spec/material-design/introduction.html
I'm torn between material design and bootstrap4 but I'll decide later once I do the routing, core HTML and client JS first.

### Feedback
* Should we be showing nodes or clients or projects? or maybe have the option for all of them?
* time intervals: Is 1, 14, 30, 90 days enough for invoicing?

**Comment by ElijahCaine** [comment]
@alxngyn Is there any reason we would do all of our visualizing client side in Javascript instead of server-side in ruby? We can do things like graphing in [Chartkick](https://github.com/ankane/chartkick), [Lazy High Charts](https://github.com/michelson/lazy_high_charts), or [gruff](https://github.com/topfunky/gruff) just to name a few I found looking around.

As far as structure of data, Since the use case is billing, we should probably allow users to start high-level (clients) and fine-grain their data from there. For instance:
1. User selects client(s) from a list
2. This triggers a visualization for aggregate billing details on this client.
3. User optionally selects a subset of projects and gets shown aggregate billing details for each project (with a breakdown below the aggregate visualization).
4. Each project can also be broken down further. I'm not sure how that will work as far as the UI goes. Maybe each project has a `nodes` dropdown with data on that project's nodes.
5. Lastly the user can `Download` a `JSON or CSV or XML` report at the top of the page.

Basically the focus should be on the data the user wants (the *forest*) and less on the individual nodes (the *trees*), and allow them to increase the granularity of the report.

I can mock this up if it's a bit confusing.

**Comment by ElijahCaine** [comment]
If we wanted to do something similar to what has already been mocked up, there can be a query builder where the user selects N clients, then M projects, and finally X nodes. Then the user gets shown a report of nodes labeled with their projects/clients and they can download a CSV or whatever of that specific subset of our data.

**Comment by alxngyn** [comment]
My programming upbringing has always been serving the data and have the client build it to off load our server's workload. I'm open to doing it either way.

You're right on the flow for the UI. I think I've been stuck in sysadmin metrics for too long.
I think this is what you mean:
```
Clients -> Projects - > Plugin -> node -> data visualization
```
So in the end the user will be able to select from a list of nodes then a list of projects then a list of plugins, and so on. The invoicing bill and data will reduce and filter down as we go down the pipeline. 

This brings up a question to the situation where if we don't know the client or project how do we work our way 'up' from the node?

**Comment by ElijahCaine** [comment]
@alxngyn To address the last point: The user can *Select All* anywhere down the pipeline and probably click to add certain Clients/Projects/etc.

**Comment by alxngyn** [comment]
Update: 
* Finish a demo next week
* make a table of disk size, ram, vcpu, disk type by fqdn

**Comment by alxngyn** [comment]
Closing this issue as conversation will go to issue #102
